### PR TITLE
4 ➡️ 5

### PR DESCRIPTION
**Pull request by chapulina**
# ➡️ Forward port

Port `ign-rendering4` to `main`

Branch comparison: https://github.com/ignitionrobotics/ign-rendering/compare/main...ign-rendering4

**Note to maintainers**: Remember to **Merge** with commit (not squash-merge or rebase)

🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

### CODE REVIEW

File: test/integration/gpu_rays.cc
Comment (chapulina) [inline test/integration/gpu_rays.cc:601]: This test is failing now:

<details>
  <summary>GpuRays/GpuRaysTest.RaysParticles</summary>

```
  [ RUN      ] GpuRays/GpuRaysTest.RaysParticles/ogre2
  /github/workspace/test/integration/gpu_rays.cc:618: Failure
  Value of: particleHit || particleMiss
    Actual: false
  Expected: true
  actual vs expected particle range: 1.3223873376846313 vs 1
  /github/workspace/test/integration/gpu_rays.cc:618: Failure
  Value of: particleHit || particleMiss
    Actual: false
  Expected: true
  actual vs expected particle range: 1.3650697469711304 vs 1
  /github/workspace/test/integration/gpu_rays.cc:618: Failure
  Value of: particleHit || particleMiss
    Actual: false
  Expected: true
  actual vs expected particle range: 1.375098705291748 vs 1
  /github/workspace/test/integration/gpu_rays.cc:618: Failure
  Value of: particleHit || particleMiss
    Actual: false
  Expected: true
  actual vs expected particle range: 1.3158919811248779 vs 1
  /github/workspace/test/integration/gpu_rays.cc:618: Failure
  Value of: particleHit || particleMiss
    Actual: false
  Expected: true
  actual vs expected particle range: 1.3633447885513306 vs 1
  /github/workspace/test/integration/gpu_rays.cc:618: Failure
  Value of: particleHit || particleMiss
    Actual: false
  Expected: true
  actual vs expected particle range: 1.3735580444335938 vs 1
  /github/workspace/test/integration/gpu_rays.cc:618: Failure
  Value of: particleHit || particleMiss
    Actual: false
  Expected: true
  actual vs expected particle range: 1.3367263078689575 vs 1
  /github/workspace/test/integration/gpu_rays.cc:618: Failure
  Value of: particleHit || particleMiss
    Actual: false
  Expected: true
  actual vs expected particle range: 1.3569036722183228 vs 1
  /github/workspace/test/integration/gpu_rays.cc:618: Failure
  Value of: particleHit || particleMiss
    Actual: false
  Expected: true
  actual vs expected particle range: 1.3486930131912231 vs 1
  /github/workspace/test/integration/gpu_rays.cc:618: Failure
  Value of: particleHit || particleMiss
    Actual: false
  Expected: true
  actual vs expected particle range: 1.3585772514343262 vs 1
  /github/workspace/test/integration/gpu_rays.cc:618: Failure
  Value of: particleHit || particleMiss
    Actual: false
  Expected: true
  actual vs expected particle range: 1.338756799697876 vs 1
  /github/workspace/test/integration/gpu_rays.cc:618: Failure
  Value of: particleHit || particleMiss
    Actual: false
  Expected: true
  actual vs expected particle range: 1.3856438398361206 vs 1
  /github/workspace/test/integration/gpu_rays.cc:618: Failure
  Value of: particleHit || particleMiss
    Actual: false
  Expected: true
  actual vs expected particle range: 1.3853719234466553 vs 1
  /github/workspace/test/integration/gpu_rays.cc:618: Failure
  Value of: particleHit || particleMiss
    Actual: false
  Expected: true
  actual vs expected particle range: 1.3508977890014648 vs 1
  /github/workspace/test/integration/gpu_rays.cc:618: Failure
  Value of: particleHit || particleMiss
    Actual: false
  Expected: true
  actual vs expected particle range: 1.3497778177261353 vs 1
  /github/workspace/test/integration/gpu_rays.cc:618: Failure
  Value of: particleHit || particleMiss
    Actual: false
  Expected: true
  actual vs expected particle range: 1.3797192573547363 vs 1
  /github/workspace/test/integration/gpu_rays.cc:618: Failure
  Value of: particleHit || particleMiss
    Actual: false
  Expected: true
  actual vs expected particle range: 1.3519102334976196 vs 1
  /github/workspace/test/integration/gpu_rays.cc:618: Failure
  Value of: particleHit || particleMiss
    Actual: false
  Expected: true
  actual vs expected particle range: 1.3346030712127686 vs 1
  [  FAILED  ] GpuRays/GpuRaysTest.RaysParticles/ogre2, where GetParam() = "ogre2" (9448 ms)
```  

</details>

Any guidance on how this should be ported forward, @iche033 ?

Comment (iche033) [reply test/integration/gpu_rays.cc:601]: ran into this test failure in https://github.com/ignitionrobotics/ign-rendering/pull/269 and should be fixed by that PR

Comment (chapulina) [reply test/integration/gpu_rays.cc:604]: #269 is now included in this PR

File: src/ParticleEmitter_TEST.cc
Comment (adlarkin) [inline src/ParticleEmitter_TEST.cc:98]: Would it be too late to make this update in Edifice?

**Review by adlarkin — COMMENTED** [review]
Windows CI failed, is that expected for Edifice (`main` branch)? I believe that Windows is usually green for Dome. Clicking on the "details" button of Windows CI leads to a 404, so I can't get any more information about the failure.

File: src/ParticleEmitter_TEST.cc
Comment (chapulina) [reply src/ParticleEmitter_TEST.cc:98]: This is in the line between bug fix and new feature. So I think it's ok to address it, but let's open a PR asap, like today!

Comment (adlarkin) [reply src/ParticleEmitter_TEST.cc:98]: Working on it now, I'll post a follow-up here when the PR is open

Comment (adlarkin) [reply src/ParticleEmitter_TEST.cc:98]: Edit: I just realized that I can't make the PR for Edifice until we merge this :upside_down_face: so, once we merge this, I will make a PR!

**Comment by chapulina** [comment]
@adlarkin , I addressed your comments and pushed, let's see if Windows CI comes back now.

**Comment by adlarkin** [comment]
Windows already failed, it looks like there's an issue with getting dependencies for the build (https://build.osrfoundation.org/job/ign_rendering-pr-win/1442/consoleFull#1273209513aa60f765-a60a-427d-900c-dc128ab22287):

```
Missing: ignition-common4 (Components: graphics, events)
```

**Comment by chapulina** [comment]
Ahhhh

`-- 	Cannot build component [graphics] - Missing: ignition-utils1`

Here we go: https://github.com/ignition-tooling/gazebodistro/pull/30

**Comment by adlarkin** [comment]
Nice catch, I missed that when scanning Windows CI. https://github.com/ignition-tooling/gazebodistro/pull/30 has been approved, so I think all we need to do now is merge that and then run Windows CI one last time :crossed_fingers:

**Comment by chapulina** [comment]
All :green_circle:

**Review by adlarkin — APPROVED** [review]
:ship: it!

**Pull request merged**
